src/test/kotlin/org/jetBrains/research/tasktracker/tracking/TaskFileHandlerTest.kt

```diff
@@ -0,0 +1,163 @@
+        val file1 = File("${project.basePath}/test1")
+        val file2 = File("${project.basePath}/test2")
+        val directory = File("${project.basePath}/tests/")
+        ApplicationManager.getApplication().runWriteAction {
+            FileUtil.createIfDoesntExist(file1)
+            FileUtil.createIfDoesntExist(file2)
+            FileUtil.createDirectory(directory)
+        }
+        val virtualFile1 = LocalFileSystem.getInstance().findFileByIoFile(file1) ?: error("file must exist")
+        val virtualFile2 = LocalFileSystem.getInstance().findFileByIoFile(file2) ?: error("file must exist")
+        val virtualDirectory =
+            LocalFileSystem.getInstance().findFileByIoFile(directory) ?: error("file must exist")
+        assert(virtualFile1.exists())
+        assert(virtualFile2.exists())
+        assert(virtualDirectory.exists())
+        val method = getPrivateFunc("addVirtualFileListener")
+        assertNoThrowable {
+            method.call(taskFileHandler, virtualFile1)
+            method.call(taskFileHandler, virtualFile2)
+            method.call(taskFileHandler, virtualDirectory)
+        }
+    }
+
+    fun testRemoveVirtualFileListener() {
+        val file = File("${project.basePath}/test")
+        val directory = File("${project.basePath}/tests/")
+        ApplicationManager.getApplication().runWriteAction {
+            FileUtil.createIfDoesntExist(file)
+            FileUtil.createDirectory(directory)
+        }
+        val virtualFile = LocalFileSystem.getInstance().findFileByIoFile(file) ?: error("file must exist")
+        val virtualDirectory =
+            LocalFileSystem.getInstance().findFileByIoFile(directory) ?: error("file must exist")
+        assert(virtualFile.exists())
+        assert(virtualDirectory.exists())
+        val methodRemove = getPrivateFunc("removeVirtualFileListener")
+        assertFailsWith<Throwable> {
+            methodRemove.call(taskFileHandler, virtualFile)
+        }
+        val methodAdd = getPrivateFunc("addVirtualFileListener")
+        assertNoThrowable {
+            methodAdd.call(taskFileHandler, virtualFile)
+            methodRemove.call(taskFileHandler, virtualFile)
+            methodRemove.call(taskFileHandler, virtualDirectory)
+        }
+        assertFailsWith<Throwable> {
+            methodRemove.call(taskFileHandler, virtualFile)
+        }
+    }
+
+    fun testGetOrCreateFile() {
+        val method = getPrivateFunc("getOrCreateFile")
+        val file1 = method.call(taskFileHandler, project, task1) as VirtualFile? ?: error("file must exist")
+        val file2 = method.call(taskFileHandler, project, task2) as VirtualFile? ?: error("file must exist")
+        assert(file1.exists())
+        assert(file2.exists())
+        assert(file1.path == "${project.basePath}/$PLUGIN_NAME/cpp/task1.cpp")
+        assert(file2.path == "${project.basePath}/$PLUGIN_NAME/tasks/task2.kt")
+        assert(String(file1.contentsToByteArray()) == task1.getContent())
```

Comment:
It is a very strange way of getting file's content....It is better to get a document or a psi file and get its text:
```kotlin
        FileDocumentManager.getInstance().getDocument(file1)?.text
        PsiManager.getInstance(project).findFile(file1)?.text
```
since the virtual file can store not only text